paddle/phi/core/generator.h

```diff
@@ -14,64 +14,109 @@ limitations under the License. */
 
 #pragma once
 
-#include <stdint.h>
-
 #include <atomic>
+#include <cstdint>
 #include <deque>
-#include <iostream>  // temp for debug
 #include <memory>
 #include <mutex>  // NOLINT
 #include <random>
 #include <typeinfo>
 #include <utility>
+#include <vector>
 
 #include "paddle/phi/common/place.h"
 
 namespace phi {
 
+#define MAGIC_RANDOM_SEED 34342423252
 class Generator {
  public:
   struct GeneratorState {
-    int64_t device = -1;
-    uint64_t current_seed = 34342423252;
-    uint64_t thread_offset = 0;
-    std::mt19937_64 cpu_engine;
+    int64_t device;
+    uint64_t seed;
+    uint64_t offset;
+    std::shared_ptr<std::mt19937_64> cpu_engine;
```

Comment:
This design choice is highlighted in the original code, which can be reviewed [here](https://github.com/PaddlePaddle/Paddle/blob/v2.6.1/paddle/phi/core/generator.cc#L180C1-L183C37). The code snippet demonstrates an interesting pattern. 

```
auto seed = GetRandomSeed();
std::seed_seq seq({seed});
auto engine = std::make_shared<std::mt19937_64>(seq);
this->state_.cpu_engine = *engine;
this->engine_ = engine;
```

An `mt19937_64` engine is allocated on the heap, and `GeneratorState` retains a copy of this initial state. However, this copy is subsequently unused. Instead, `Generator` has a `shared_ptr` pointing to the `mt19937_64` engine on the heap, and this pointer is used for engine manipulation, as seen in [`GetCPUEngine`](https://github.com/PaddlePaddle/Paddle/blob/v2.6.1/paddle/phi/core/generator.cc#L263) and [`Random64`](https://github.com/PaddlePaddle/Paddle/blob/v2.6.1/paddle/phi/core/generator.cc#L273).

The original code seems to favor managing the `mt19937_64` engine on the heap by the shared_ptr rather than the engine held by `GeneratorState`. If `GeneratorState` were to own the `mt19937_64` engine directly, a shared_ptr would not be necessary since the engine's lifetime would align with the state's lifetime. In this case, a normal pointer would suffice.

I have considered letting `GeneratorState` directly own the `mt19937_64` engine. However, this would entail changing the `GetCPUEngine` API to a normal pointer, potentially impacting numerous files that utilize this API.